cmd/gonix/nar/ls.go

```diff
@@ -0,0 +1,89 @@
+package nar
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/nix-community/go-nix/pkg/nar"
+)
+
+type LsCmd struct {
+	Nar       string `kong:"cmd,arg,type:'existingfile'"`
+	Path      string `kong:"cmd,arg,optional,type:'string',default:'',help:'Path inside the NAR to filter'"`
+	Recursive bool   `kong:"cmd,short:'R',help:'Whether to list recursively, or only the current level.'"`
+}
+
+// headerLineString returns a one-line string describing a header.
+// hdr.Validate() is assumed to be true.
+func headerLineString(hdr *nar.Header) string {
+	var sb strings.Builder
+
+	sb.WriteString(hdr.FileInfo().Mode().String())
+	sb.WriteString(" ")
+
+	// write name. In case of an empty path, write a "."
+	if hdr.Path == "" {
+		sb.WriteString(".")
+	} else {
+		sb.WriteString(hdr.Path)
+	}
+
+	// if regular file, show size in parantheses. We don't bother about aligning it nicely,
+	// as that'd require reading in all headers first before printing them out.
+	if hdr.Size > 0 {
+		sb.WriteString(fmt.Sprintf(" (%v bytes)", hdr.Size))
+	}
+
+	// if LinkTarget, show it
+	if hdr.LinkTarget != "" {
+		sb.WriteString(" -> ")
+		sb.WriteString(hdr.LinkTarget)
+	}
+
+	sb.WriteString("\n")
+
+	return sb.String()
+}
+
+func (cmd *LsCmd) Run() error {
+	f, err := os.Open(cmd.Nar)
+	if err != nil {
+		return err
+	}
+
+	nr, err := nar.NewReader(f)
+	if err != nil {
+		return err
+	}
+
+	for {
+		hdr, err := nr.Next()
+		if err != nil {
+			// io.EOF means we're done
+			if err == io.EOF {
+				return nil
+			}
+			// relay other errors
+			return err
+		}
+
+		// if the yielded path starts with the path specified
+		if strings.HasPrefix(hdr.Path, cmd.Path) {
```

Comment:
`hdr.Path` will miss the leading `/` here, preventing `gonix ls $nar /(.*)` to match anything.